src/controllers/auth.controller.js

```diff
@@ -0,0 +1,269 @@
+  const userData = await jwtService.verifyRefresh(refreshToken);
+
+  if (!userData || !refreshToken) {
+    throw ApiError.unauthorized();
+  }
+
+  await tokenService.remove(userData.id);
+
+  res.sendStatus(204);
+};
+
+const passwordReset = async (req, res) => {
+  const { email } = req.body;
+  const user = await userService.findByEmail(email);
+
+  if (!user) {
+    throw ApiError.unauthorized('User with that email does not exist');
+  }
+
+  const resetToken = jwtService.sign(userService.normalize(user));
+
+  await tokenService.save(user.id, resetToken);
+  await sendResetEmail.sendResetEmail(user.email, resetToken);
+
+  res.send({ message: 'Password reset link sent to your email' });
+};
+
+const resetPassword = async (req, res) => {
+  const { token, password, confirmPassword } = req.body;
+
+  if (password !== confirmPassword) {
+    throw ApiError.badRequest('Passwords do not match');
+  }
+
+  const userData = jwtService.verify(token);
+
+  if (!userData) {
+    throw ApiError.badRequest('Invalid or expired token');
+  }
+
+  const user = await User.findByPk(userData.id);
+
+  if (!user) {
+    throw ApiError.badRequest('User not found');
+  }
+
+  const hashedPass = await bcrypt.hash(password, 10);
+
+  user.password = hashedPass;
+
+  await user.save();
+
+  res.send({ message: 'Password successfully reset' });
+};
+
+const updateProfile = async (req, res) => {
+  const {
+    name,
+    oldPassword,
+    newPassword,
+    confirmPassword,
+    newEmail,
+    confirmEmail,
+  } = req.body;
+  const userId = req.user.id;
+
+  const user = await User.findByPk(userId);
+
+  if (!user) {
+    throw ApiError.notFound('User not found');
+  }
+
+  if (name) {
+    if (name.trim().length < 2) {
+      throw ApiError.badRequest('Name should be at least 2 characters long');
+    }
+
+    user.name = name.trim();
+  }
+
+  if (newPassword) {
+    if (!oldPassword) {
+      throw ApiError.badRequest('Old password is required');
+    }
+
+    if (newPassword !== confirmPassword) {
+      throw ApiError.badRequest('New password and confirmation do not match');
+    }
+
+    const isValidPassword = await bcrypt.compare(oldPassword, user.password);
+
+    if (!isValidPassword) {
+      throw ApiError.badRequest('Invalid old password');
+    }
+
+    const hashedPassword = await bcrypt.hash(newPassword, 10);
+
+    user.password = hashedPassword;
+  }
+
+  if (newEmail) {
+    if (newEmail !== confirmEmail) {
+      throw ApiError.badRequest('New email and confirmation do not match');
+    }
+
+    if (!oldPassword) {
+      throw ApiError.badRequest('Old password is required for email change');
+    }
+
+    const isValidPassword = await bcrypt.compare(oldPassword, user.password);
+
+    if (!isValidPassword) {
+      throw ApiError.badRequest('Invalid old password for email change');
+    }
+
+    const emailError = validateEmail(newEmail);
+
+    if (emailError) {
+      throw ApiError.badRequest('Invalid email', { email: emailError });
```

Comment:
Consider using `return next(ApiError.badRequest('Invalid email', { email: emailError }));` instead of `throw` to ensure the error is passed to the next middleware for consistent error handling.